src/analysis/functions/mod.rs

```diff
@@ -0,0 +1,159 @@
+/// allows for generic filtering to remove un-important arguments.
+pub trait ArgumentFilter {
+    fn is_important(ssa: &SSAStorage, ni: NodeIndex) -> bool;
+}
+
+/// Let's every argument through, no filtering applies
+struct NoArgumentFilter;
+impl ArgumentFilter for NoArgumentFilter {
+    fn is_important(_ssa: &SSAStorage, _ni: NodeIndex) -> bool {
+        true
+    }
+}
+
+struct X86ArgumentFilter;
+impl ArgumentFilter for X86ArgumentFilter {
+    fn is_important(ssa: &SSAStorage, ni: NodeIndex) -> bool {
+        let nd = ssa.node_data(ni).unwrap();
+        let operands = ssa.operands_of(ni);
+
+        let dst_registers = ssa.registers(ni);
+
+        // rax is conventionally used as at the return value register
+        if dst_registers.len() == 1 && dst_registers[0] == "rax" {
+            return false;
+        }
+
+        match nd.nt {
+            NodeType::Op(ref opc) => {
+                match opc {
+                    &MOpcode::OpLoad => {
+                        // Ignore anything that loads a value into rip, this is usually
+                        // the "ret" instruction popping the return address from the stack
+                        // and we don't treat the return address as an explicit argument
+                        if dst_registers.len() == 1 && dst_registers[0] == "rip" {
+                            return false;
+                        }
+                    },
+                    &MOpcode::OpAdd | &MOpcode::OpSub => {
+                        // Ignore instructions that extend or shrink the stack, these look
+                        // like rsp = rsp + 0x32 or rsp = rsp - 0x8 etc. While technically
+                        // the original stack pointer is an argument to the function, we
+                        // will not consider it to be one.
+                        if dst_registers.len() == 1 && dst_registers[0] == "rsp" {
+                            return false;
+                        }
+                    }
+                    _ => ()
+                };
+            },
+            _ => ()
+        }
+
+        true
+    }
+}
+
+impl<'a> FuncArgumentAnalyzer<'a> {
+    pub fn new(registers: &'a SubRegisterFile, sections: &'a [LSectionInfo]) -> FuncArgumentAnalyzer<'a> {
+        return FuncArgumentAnalyzer {
+            registers, sections, arguments: Vec::new()
+        };
+    }
+
+    pub fn determine_arguments(&self, f: &mut RadecoFunction) -> FuncInfo {
+        self.determine_arguments_filter::<X86ArgumentFilter>(f)
+    }
+
+    pub fn determine_arguments_filter<T: ArgumentFilter>(&self, f: &mut RadecoFunction) -> FuncInfo {
+        println!("Parsing new function: {}", f.name);
+
+        if f.name != "sym._init" {
+            let mut arguments = Vec::new();
+            return FuncInfo { arguments };
+        }
+
+        let ssa = f.ssa();
+
+        let mut arguments = Vec::new();
+        let mut tainted = HashSet::new();
+
+        for idx in ssa.inorder_walk() {
```

Comment:
There's a simpler way to implement the above. Let me try and explain it.

The entry block of every function is a dummy block that provides definitions for every register and memory that is a part of the architecture. Therefore, what you could do is, iterate over these definitions in the entry block, and collect all its uses. This would give you all the registers which are *use before def* (and a subset of these would be arguments). As you've outlined, there would be false positives which arise due to the cases that you've mentioned and also due to instructions which are used to *save* register values (such as `push <reg>`, and these would be use before def for `<reg>` as well).

Reply:
This works because if you had an instruction that defined a register before using it, such as `mov <reg>, <expr>`, then the IR construction would use the `<expr>` node rather than the define from the start block.